code/Test_definitions/Location-Retrieval.feature

```diff
@@ -0,0 +1,264 @@
+    When the HTTP "POST" request is sent
+    Then the response status code is 422
+    And the response property "$.status" is 422
+    And the response property "$.code" is "DEVICE_IDENTIFIERS_MISMATCH"
+    And the response property "$.message" contains a user friendly text
+
+  @location_retrieval_15_device_token_mismatch
+  Scenario: Inconsistent access token context for the device
+    # To test this, a token have to be obtained for a different device
+    Given the request body property "$.device" is set to a valid testing device
+    And the header "Authorization" is set to a valid access token emitted for a different device
+    When the HTTP "POST" request is sent
+    Then the response status code is 403
+    And the response property "$.status" is 403
+    And the response property "$.code" is "INVALID_TOKEN_CONTEXT"
+    And the response property "$.message" contains a user friendly text
+
+  @location_retrieval_16_device_not_supported
+  Scenario: Service not available for the device
+    Given that service is not supported for all devices commercialized by the operator 
+    And the request body property "$.device" is set to a valid device for which the service is not applicable
+    When the HTTP "POST" request is sent
+    Then the response status code is 422
+    And the response property "$.status" is 422
+    And the response property "$.code" is "DEVICE_NOT_APPLICABLE"
+    And the response property "$.message" contains a user friendly text
+
+  # Generic 400 errors
+
+  @location_retrieval_400.1_no_request_body
+  Scenario: Missing request body
+    Given the request body is not included
+    When the HTTP "POST" request is sent
+    Then the response status code is 400
+    And the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+
+  @location_retrieval_400.2_empty_request_body
+  Scenario: Empty object as request body
+    Given the request body is set to "{}"
+    When the HTTP "POST" request is sent
+    Then the response status code is 400
+    And the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+
+  # Other specific 400 errors
+
+  @location_retrieval_400.3_max_age_schema_compliant
+  Scenario: Input property values doe not comply with the schema
+    Given the request body property "$.device" is set to a valid testing device
+    And the "maxAge" is set to 6a0
+    When the HTTP "POST" request is sent
+    Then the response status code is 400
+    And the response property "$.status" is 400
+    And the response property "$.code" is "INVALID_ARGUMENT"
+    And the response property "$.message" contains a user friendly text
+
+
+  @location_retrieval_400.4_required_device_identifier_missing

```

Comment:
Will device remain a mandatory field in the next release?
(Referring to three legged use case where access token is tied to a device)

Reply:
This is a fair question.
By the generic rule this API will require 3-legs but what if this API is used to track device (not linked to people) for a B2B customer? I tend to think we can have some borderline UC in 2-legs. I prefer to keep it as optional but if we prefer to remove it I understand.
@jlurien WDYT?

Reply:
If we make device optional this test case would need to be adapted. For 2-legged access,  or even for 3-legged when a single device cannot be deducted from the access token, it would be an error, but not 400 INVALID_ARGUMENT (not compliant with schema)  but 422 UNIDENTIFIABLE_DEVICE